lib/spack/spack/util/web.py

```diff
@@ -665,6 +663,11 @@ def find_versions_of_archive(
             except spack.url.UndetectableVersionError:
                 continue
 
+    for url in archive_urls:
+        ver = spack.url.parse_version(url)
+        if ver not in versions:
+            versions[ver] = url
```

Comment:
(question) I think placing this logic here inverts the original prioritization - is that intended?

i.e. originally the `archive_urls` would take precedence (because of `if ver in matched: continue`), but now they are only used if the links are not.

Reply:
Yes, it's very much intentional, that's to fix one of the explicit bugs.  The issue is that sometimes the URLs from the package aren't known to be correct, so prioritizing the archive_urls was pinning non-functional links without a way to tell.  Now it prioritizes a detected link over the archive URL on the assumption that if it found it that way it's more likely to be valid.  I would worry about this causing issues if this function were always used first, but it isn't, it's normally used as a fallback after the archive URL or URLs have been tried.

Reply:
That sounds good: I think it would be worthwhile to put this explanation in the docstring (or rather to explain what gets the highest priority, not necessarily to delve into why).